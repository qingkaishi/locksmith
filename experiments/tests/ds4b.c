/* Data structure example 4:  A recursive data structure.  
   The list elements themselves are protected, but what
   they point to is not. */

#include <pthread.h>

struct list {
  int data;
  struct list *next;
};

pthread_mutex_t global_lock;
struct list *shared = NULL;

void *run(void *arg) {
  struct list *cur;

  pthread_mutex_lock(&global_lock);
  for (cur = shared; cur; cur = cur->next) {
    int *data = &cur->data;
    pthread_mutex_unlock(&global_lock);
    (*data)++; // Race!
    pthread_mutex_lock(&global_lock);
  }
  pthread_mutex_unlock(&global_lock);

  return NULL;
}
int main() {
  pthread_t t1, t2;
  int i;

  for (i = 0; i < 42; i++) {
    struct list *new = (struct list *) malloc(sizeof(struct list));
    new->next = shared;
    shared = new;
  }
  pthread_mutex_init(&global_lock, NULL);
  pthread_create(&t1, NULL, run, NULL);
  pthread_create(&t2, NULL, run, NULL);
  return 1;
}
