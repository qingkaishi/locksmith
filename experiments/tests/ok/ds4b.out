************************* STARTING *************************

shared: alloc.data:ds4b.c:34

Warning: Possible data race: alloc.data:ds4b.c:34 is not protected!
 references:
  dereference of data:ds4b.c:20 at ds4b.c:22
    alloc.data:ds4b.c:34 => *cur.data:ds4b.c:13
                         => data:ds4b.c:20
  locks acquired:
    <empty>
  in: main at ds4b.c:29 


*************************** DONE ***************************
